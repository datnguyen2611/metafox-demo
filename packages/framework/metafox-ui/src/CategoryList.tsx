import { Link, useGlobal } from '@metafox/framework';
import { DotSeparator } from '@metafox/ui';
import { SxProps } from '@mui/material';
import React from 'react';

interface Item {
  name: string;
  id: number;
}

interface CategoryListProps<T extends Item = Item> {
  data: T[];
  className?: string;
  sx?: SxProps;
  isTruncateOneLine?: boolean;
  displayLimit?: number;
}

export function CategoryList({
  data,
  sx,
  isTruncateOneLine = false,
  displayLimit,
  ...props
}: CategoryListProps) {
  const { i18n } = useGlobal();
  const [categoryList, setCategoryList] = React.useState(
    data?.length ? data : []
  );

  React.useEffect(() => {
    if (displayLimit && data) {
      const dataTmp: any[] = data.slice(0, displayLimit);

      if (data.length > displayLimit) {
        dataTmp.push({
          id: 0,
          name: i18n.formatMessage(
            { id: 'total_more_categories' },
            {
              total: data.length - displayLimit
            }
          )
        });
      }

      setCategoryList(dataTmp);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [displayLimit, data]);

  if (!data?.length) return null;

  const handleShowMore = () => {
    setCategoryList(data);
  };

  return (
    <DotSeparator
      sx={{
        textTransform: 'uppercase',
        fontSize: '13px',
        display: 'flex !important',
        flexWrap: 'wrap',
        lineHeight: 1,
        color: 'primary.main',
        ...sx
      }}
      isTruncateOneLine={isTruncateOneLine}
      {...props}
    >
      {categoryList.map((item, index) => {
        if (item.id === 0) {
          return (
            <Link key={'more'} onClick={handleShowMore}>
              {item.name}
            </Link>
          );
        }

        return (
          <Link key={index.toString()} to={item?.url}>
            {item.name}
          </Link>
        );
      })}
    </DotSeparator>
  );
}
