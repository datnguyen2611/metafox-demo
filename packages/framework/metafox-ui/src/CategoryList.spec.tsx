import * as React from 'react';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';
import { CategoryList } from './CategoryList';

describe('CategoryList', () => {
  it('render null', async () => {
    const props = {
      data: [],
      to: ''
    };
    const tree = renderer.create(<CategoryList {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('render item', async () => {
    const props = {
      data: [
        { id: 1, name: 'category 1' },
        { id: 2, name: 'category 2' }
      ],
      to: '/blog'
    };
    const tree = renderer
      .create(
        <MemoryRouter>
          <CategoryList {...props} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
