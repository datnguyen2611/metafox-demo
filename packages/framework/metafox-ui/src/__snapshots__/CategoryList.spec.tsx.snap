// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList render item 1`] = `
<div
  className="MuiBox-root css-1yy5cbc-DotSeparator-root"
>
  <span>
    <a
      className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
      href="/blog/1"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
    >
      category 1
    </a>
  </span>
  <span>
    <a
      className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
      href="/blog/2"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
    >
      category 2
    </a>
  </span>
</div>
`;

exports[`CategoryList render null 1`] = `null`;
